internal/statestore/redis.go

```diff
@@ -20,29 +20,48 @@ import (
 	"io/ioutil"
 	"time"
 
-	"github.com/cenkalti/backoff"
-	"github.com/golang/protobuf/proto"
-	"github.com/gomodule/redigo/redis"
-	"github.com/pkg/errors"
 	"github.com/sirupsen/logrus"
 	"google.golang.org/grpc/codes"
 	"google.golang.org/grpc/status"
 	"open-match.dev/open-match/internal/config"
-	"open-match.dev/open-match/pkg/pb"
-)
 
-const (
-	allTickets        = "allTickets"
-	proposedTicketIDs = "proposed_ticket_ids"
+	rs "github.com/go-redsync/redsync/v4"
+	rsredigo "github.com/go-redsync/redsync/v4/redis/redigo"
+	"github.com/gomodule/redigo/redis"
 )
 
 var (
 	redisLogger = logrus.WithFields(logrus.Fields{
 		"app":       "openmatch",
 		"component": "statestore.redis",
 	})
+
+	// this field is used to create new mutexes
+	redsync *rs.Redsync
 )
 
+// redisLocker is a wrapper over redsync mutex in order to let callers use it but not to expose redsync package outside
+type redisLocker struct {
+	mutex *rs.Mutex
+}
+
+// NewMutex returns a new distributed mutex with given name
+func (rb *redisBackend) NewMutex(key string) *redisLocker {
+	//TODO: make expiry duration configurable
+	m := redsync.NewMutex(key, rs.WithExpiry(5*time.Minute))
+	return &redisLocker{mutex: m}
+}
+
+// Lock locks r. In case it returns an error on failure, you may retry to acquire the lock by calling this method again.
+func (r redisLocker) Lock() error {
+	return r.mutex.Lock()
+}
```

Comment:
While performing E2E tests on Update Backfill, found that context is needed here:
```suggestion
// Lock locks r. In case it returns an error on failure, you may retry to acquire the lock by calling this method again.
func (r redisLocker) Lock(ctx context.Context) error {
	return r.mutex.LockContext(ctx)
}
```
Otherwise nil pointer dereference in `func GetRedisPool(cfg config.View) *redis.Pool {` -`dialFunc`:
```
			if ctx.Err() != nil 
```

Reply:
Nil pointer exceptionon appears because there is no `err != nil` check in dialFunc, but it should be there. I will add context anyway.